examples/chef/common/chef-fan-control-manager.cpp

```diff
@@ -99,98 +112,223 @@ Status ChefFanControlManager::HandleStep(StepDirectionEnum aDirection, bool aWra
+    ChipLogProgress(NotSpecified, "ChefFanControlManager::HandleFanControlAttributeChange");
+    switch (attributeId)
     {
-    case SpeedSetting::Id: {
-        Nullable<uint8_t> newSpeedSetting;
-        ReturnErrorOnFailure(aDecoder.Decode(newSpeedSetting));
-
-        // Ensure new speed is in bounds
+    case FanControl::Attributes::PercentSetting::Id: {
+        ChipLogProgress(NotSpecified, "ChefFanControlManager::HandleFanControlAttributeChange  PercentSetting");
+        DataModel::Nullable<Percent> percentSetting = static_cast<DataModel::Nullable<uint8_t>>(*value);
+        if (percentSetting.IsNull())
         {
-            uint8_t maxSpeedSetting                    = 0;
-            Protocols::InteractionModel::Status status = SpeedMax::Get(mEndpoint, &maxSpeedSetting);
-            VerifyOrReturnError(status == Protocols::InteractionModel::Status::Success, CHIP_IM_GLOBAL_STATUS(Failure));
-
-            if (!newSpeedSetting.IsNull() && newSpeedSetting.Value() > maxSpeedSetting)
-            {
-                return CHIP_IM_GLOBAL_STATUS(ConstraintError);
-            }
+            PercentSettingWriteCallback(0);
         }
-
-        // Only act on changed.
-        if (newSpeedSetting != mSpeedSetting)
+        else
         {
-            mSpeedSetting = newSpeedSetting;
+            PercentSettingWriteCallback(percentSetting.Value());
+        }
+        break;
+    }
 
-            // Mark both the setting AND the current dirty, since the current always
-            // tracks the target for our product.
-            MatterReportingAttributeChangeCallback(mEndpoint, FanControl::Id, Attributes::SpeedSetting::Id);
-            MatterReportingAttributeChangeCallback(mEndpoint, FanControl::Id, Attributes::SpeedCurrent::Id);
+    case FanControl::Attributes::SpeedSetting::Id: {
+        ChipLogProgress(NotSpecified, "ChefFanControlManager::HandleFanControlAttributeChange  SpeedSetting");
+
+        DataModel::Nullable<uint8_t> speedSetting = static_cast<DataModel::Nullable<uint8_t>>(*value);
+        if (speedSetting.IsNull())
+        {
+            SpeedSettingWriteCallback(0);
         }
+        else
+        {
+            SpeedSettingWriteCallback(speedSetting.Value());
+        }
+        break;
+    }
+
+    case FanControl::Attributes::FanMode::Id: {
+        ChipLogProgress(NotSpecified, "ChefFanControlManager::HandleFanControlAttributeChange  FanMode");
 
+        FanControl::FanModeEnum fanMode = static_cast<FanControl::FanModeEnum>(*value);
+        FanModeWriteCallback(fanMode);
         break;
     }
-    case PercentSetting::Id: {
-        Nullable<uint8_t> newPercentSetting;
-        ReturnErrorOnFailure(aDecoder.Decode(newPercentSetting));
 
-        // Ensure new speed in percent is valid.
-        if (!newPercentSetting.IsNull() && newPercentSetting.Value() > 100)
+    default: {
+        break;
+    }
+    }
+}
+
+void ChefFanControlManager::PercentSettingWriteCallback(uint8_t aNewPercentSetting)
```

Comment:
> Looking at the spec, none of this code should be reached if null is written (cluster should ignore the write and return with an error per spec), no?

It seems this is only true when the `client` is trying to write to null.
The cluster code can still set this to null and we use the post attribute change callback to detect the change in PercentSetting so we can update PercentCurrent.

See: https://github.com/project-chip/connectedhomeip/blob/master/src/app/clusters/fan-control-server/fan-control-server.cpp#L232

Reply:
Yeah, the spec doesn't specify what to do with `PercentCurrent` when `PercentSetting` goes to null (a.k.a when the FanMode goes to kAuto).
We're deciding to set this to 0 in the app (as `PercentCurrent` can't be null as well).

Reply:
@soares-sergio The right behavior is that in auto mode PercentCurrent should reflect the actual speed of the fan, as a percentage of max, no?

Reply:
Yes!

In a real application, I would expect the app to keep updating `SpeedCurrent` and `PercentCurrent` to the real Fan speed/percentage while in auto mode.
For this example app, we simplified to writing to a fixed value (0, could be any other valid value).

A more sophisticated approach for this Example could be to build an "Auto" mode simulator that's enabled when FanMode is set to `Auto` and then automatically update `SpeedCurrent` and `PercentCurrent` (maybe keep looping through a few speeds?).

Is that what you're suggesting?

Reply:
> For this example app, we simplified to writing to a fixed value (0, could be any other valid value).

Ah, I see.  I think that's fine but should be documented, because it looked like "it's null so we don't know what to do" vs "it's null and this is the thing we are doing for this reason".

Reply:
Cool. Added a comment in the code, where we set those to 0.
```
        // The cluster code in fan-control-server.cpp is the only one allowed to set speedSetting to null.
        // This happens as a consequence of setting the FanMode to kAuto. In auto mode, speedCurrent should continue to report the
        // real fan speed. In this example, we set SpeedCurrent to 0 here as we don't have a real value for the Fan speed or a
        // FanAutoMode simulator.
        // When not Null, SpeedCurrent tracks SpeedSetting's value.
```

I'll add an AutoMode simulator in a future PR. I think it's a good idea anyway to provide an example for that. Thanks!